### ISSUE TITLE
How to link against objc categories in a static library.

### ISSUE THREAD

**Issue by rupertdaniel** [Microsoft/WinObjC#236]
I have some objc categories included in a .a static library. When I come to link the final .exe against the .a I get a bunch of linker errors like the following, and at runtime the categories are not available.

`PMKPromise+When_FD12DC43.obj : warning LNK4221: This object file does not define any previously undefined public symbols, so it will not be used by any link operation that consumes this library`

In xcode you usually define the `-load_all` param in the Other Linker Flags for this situation.

Is there an equivalent option for the VS linker?

At the moment my workaround is to build the category source .m files in the main app project.

**Comment by rajsesh-msft** [comment]
The linker has a new flag (it has had it for a while now) /wholearchive that would have supposedly fixed this problem.  However, our attempt to use it uncovered other issues (see PR #684, #874, #879). 

There are two workarounds: 
1. Define the categories in the app project (per @rupertdaniel).
2. define a dummy "C" function in the .m file that implements the extension.  Call the dummy function from app initialization code (any file in the app project).

2 is what we have been using for some internal apps that ran into this issue.  

We will revisit #684 in the future.

**Comment by edvv** [comment]
@rajsesh-msft Is workaround (2) needed only for .m files that have only a category definition; or is workaround (2) also needed for .m files that have class definitions (and perhaps also a category definition)?

**Comment by rajsesh-msft** [comment]
Only for categories and extensions. New class definitions will be picked up by the linker.

**Comment by edvv** [comment]
@rajsesh-msft What if the new class (defined in the static library) is never referenced by the app [at link time]? - like when will occur if it is looked up and bound at runtime.

**Comment by rajsesh-msft** [comment]
It should still work although I can't say I have tried. The issue here is not that the symbol is not referenced. It is simply that Class1 and Class1 (category) are the same symbols, so the new definitions are ignored (as noted by the linker error).

**Comment by DHowett-MSFT** [comment]
Nah; in this case, it _is_ because nobody is referencing the symbol. If you emit an Objective-C class that only has ` +load`, or is looked up dynamically, the linker may strip its containing object from the resulting binary. You have to anchor at least one symbol from the translation unit containing it, either with a linker include `#pragma` or by referring to the symbol in another non-stripped TU.

Categories are the same, except they have no linkable symbols at all. One must be added.

**Comment by edvv** [comment]
@DHowett-MSFT Can you please elaborate on the linker include `#pragma`, how to use it and an example?

**Comment by DHowett-MSFT** [comment]
@edvv My apologies; it looks like `#pragma comment(linker, "/include:...")` doesn't work in the static library scenario. You'll have to use an anchoring symbol like a no-op function or global referenced from a consumed TU.

My original suggestion follows, for posterity:

### Attempt
```objc
@interface X
...
@end

@implementation X
...
@end

#pragma comment(linker, "/include:__OBJC_CLASS_X")
```

Unfortunately, it relies on the ABI-mangled name of the class, as well as emitting a directive to the linker.

### Failure
```
E:\src\WinObjC\tests\unittests\Foundation\Tests.mm(43):
error: Expected: (((Class)nullptr)) != (objc_getClass("X")), actual: NULL vs NULL
```

**Comment by edvv** [comment]
@DHowett-MSFT Just a quick FYI here (referencing https://github.com/Microsoft/WinObjC/issues/1589):

This is the same way the Mach linker works on darwin (Mac and iOS), the -load_all was added about a decade ago and was always thought to be a temporary patch (from what I recall regarding discussions about it). It was supplanted by -force_load (which loads on a per archive basis). Before -load_all, static archives needed a manual reference as explained here. The linker include `#pragma`, if it worked, would have been a very nice feature and easy to use.

I see this is due to be fixed in 1704, if that can be moved up to 1702 then it would help.

Sincerely,

Ed
